lib/jxl/render_pipeline/stage_write.cc

```diff
@@ -334,17 +339,36 @@ class WriteToPixelCallbackStage : public RenderPipelineStage {
           temp[j++] = line_buffers[c][ix];
         }
       }
+      size_t xstart = xpos + x0;
+      size_t xlen = ix;
       if (has_alpha_ && want_alpha_ && unpremul_alpha_) {
         // TODO(szabadka) SIMDify (possibly in a separate pipeline stage).
-        UnpremultiplyAlpha(temp, num_color_, ix);
+        UnpremultiplyAlpha(temp, num_color_, xlen);
       }
       if (swap_endianness_) {
-        size_t len = ix * num_channels_;
+        size_t len = xlen * num_channels_;
         for (size_t j = 0; j < len; ++j) {
           temp[j] = BSwapFloat(temp[j]);
         }
       }
-      pixel_callback_.run(run_opaque_, thread_id, xpos + x0, ypos, ix, temp);
+      if (flip_x_) {
+        size_t last = (xlen - 1u) * num_channels_;
+        size_t num = (xlen / 2) * num_channels_;
+        for (size_t i = 0; i < num; i += num_channels_) {
+          for (size_t c = 0; c < num_channels_; ++c) {
+            std::swap(temp[i + c], temp[last - i + c]);
+          }
+        }
+        xstart = width_ - xstart - xlen;
+      }
+      if (transpose_) {
+        for (size_t i = 0, j = 0; i < xlen; ++i, j += num_channels_) {
+          pixel_callback_.run(run_opaque_, thread_id, ypos, xstart + i, 1,
```

Comment:
While this works, it is *rather slow*. Maybe we should (eventually) think of faster ways to do it.

Reply:
There are some tricks to do NxN transpositions with SIMD, so maybe we could buffer some lines here, and output N pixels at each call. (But transposition is an inherently slow operation and the implementation we have currently does not seem to be much better.)

Reply:
Buffering lines wouldn't necessarily work unless we keep large buffers, it is probably better to buffer i.e. 8x8 chunks and guarantee in the RP that dimensions satisfy this-and-that constraint... Anyway, we don't need to do this now, just adding a TODO is fine.

Reply:
added the TODO